ENH: stats.ecdf: add confidence intervals

mdhaber opened the pull request:
#### Reference issue
Follow-up to gh-18092

#### What does this implement/fix?
This adds confidence intervals for the CDF/SF estimated by `stats.ecdf`.

```python3
import numpy as np
from scipy import stats
import matplotlib.pyplot as plt

# Enter censored data
times = [24, 3, 11, 19, 24, 13, 14, 2, 18, 17, 24, 21, 12, 1, 10, 23, 6, 5, 9, 17]
died = [0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0, 1, 0, 1]
sample = stats.CensoredData.right_censored(times, np.logical_not(died))

# Generate point estimate and confidence interval of the survival function
res = stats.ecdf(sample)
point_estimate = res.sf.points
ci = res.sf.confidence_interval(0.90)

# plotting
ax = plt.subplot()
x = np.insert(res.x, 0, 0)  # plots look better when a leftmost point is added
ax.step(x, np.insert(point_estimate, 0, 1), color='C0', where='post')
ax.step(x, np.insert(ci.low, 0, 1), color='C0', linestyle='dashed', where='post')
ax.step(x, np.insert(ci.high, 0, 1),  color='C0', linestyle='dashed', where='post')
ax.set_xlabel("time")
ax.set_ylabel("survival function")
ax.set_title("Survival function point estimate and 90% CI")
plt.show()
```
![image](https://user-images.githubusercontent.com/6570539/224517524-c029f69d-8036-4c72-b40d-c0476b1da3eb.png)


#### Additional information
Other confidence interval methods can/will be added in follow-up PRs. The one implemented here is the oldest and most common, and it's the one Matlab uses.

I didn't add an example because the existing data yields NaNs in the variance calculation (correctly). I plan to substantially revise the example in a follow-up, at which point I'll show how to use the `confidence_interval` method.

Because this makes the `cdf` and `sf` attributes objects (instead of arrays), these objects can easily be endowed with `__call__` methods that evaluate the ECDF/ESF at user-specified points. This will respond to requests from the original PR. Perhaps we should remove the `points` attribute, after all, and the user can evaluate the functions at the sample points if desired.

It's weird that `stats.ecdf` returns an object with an attribute `sf`. It would make more sense to rename `stats.ecdf` to something like `stats.nonparametric_fit`, and `ECDFResult` would become `NonparametricFitResult`. Ultimately, these classes should be documented on their own like [`FitResult`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats._result_classes.FitResult.html#scipy.stats._result_classes.FitResult) is.

Matlab returns confidence intervals in which the first value as NaN, even though that is not produced by Greenwood's formula. Intuitively, this makes some sense, but I'd like to check against some other references before I implement this. Also, Matlab returns NaNs in the confidence intervals, but arguably `[0, 1]` could be used instead of `[np.nan, np.nan]`.

mdhaber commented on scipy/stats/_survival.py:
```diff
@@ -155,7 +303,10 @@ def _ecdf_uncensored(sample):
     sf = 1 - cdf
 
-    return ECDFResult(x, cdf, sf)
+    _n = sf*sample.size
+    _d = counts
+
+    return x, cdf, sf, _n, _d
```
```suggestion
    return x, cdf, sf, sf*sample.size, counts
```

tupui commented:
About removing ˋpoints`, I think we can consider that if E also remove `x`. In the end it's a matter of thinking how this is going to be used the most I suppose.

We could still have this private for a plot method.

About the renaming. We could although I am not sure about the name. This is quite obscure for non experts I think. Not that I am reference, but I would not have guess the meaning of that function.

Maybe `ecdf` some ok if instead we call `sf`: `ccdf`. I know sf is for consistency and linked also to survival analysis, but ccdf is more descriptive.

tupui commented on scipy/stats/_survival.py:
```diff
@@ -1,38 +1,213 @@
+        confidence_level : float, default: 0.95
+            Confidence level for the computed confidence interval
+
+        method : str, optional
```
Above you use `default`. We should do that here as well or change above to make it consistent.

mdhaber replied:
Yup.
```suggestion
        method : str, default: "Greenwood"
```

tupui commented on scipy/stats/_survival.py:
```diff
@@ -1,38 +1,213 @@
+
+        Returns
+        -------
+        ci : ``ConfidenceInterval``
```
Why the double quotes? You don't want to have the link?

mdhaber replied:
Following the precedent of `binomtest` and other functions that use this class.
![image](https://user-images.githubusercontent.com/6570539/224869058-c6adbf6e-fc53-4ac7-b2b6-dd78e8ea5672.png)

It's not publicly documented, actually, so there's nothing to link to right now.

tupui commented on scipy/stats/_survival.py:
```diff
@@ -1,38 +1,213 @@
+
+        References
+        ----------
+        .. [1] Sawyer, Stanley. "The Greenwood and Exponential Greenwood
+               Confidence Intervals in Survival Analysis."
+               https://www.math.wustl.edu/~sawyer/handouts/greenwood.pdf
```
They don't have a paper? Looks a bit weak to implement a method if there is just a PDF report otherwise. Unless we are confident this is indeed better.

tupui commented on scipy/stats/_survival.py:
```diff
@@ -1,38 +1,213 @@
+        if method.lower() not in methods:
+            raise ValueError(message)
+
+        message = "`confidence_level` must be a scalar between 0 and 1."
+        confidence_level = np.asarray(confidence_level)[()]
+        if confidence_level.shape or not (0 <= confidence_level <= 1):
+            raise ValueError(message)
```
I think we should think about a general way to check that. We are duplicating a lot of checks for that. (Not for this PR)

tupui commented on scipy/stats/_survival.py:
```diff
@@ -1,38 +1,213 @@
+        with np.errstate(divide='ignore', invalid='ignore'):
+            var = sf ** 2 * np.cumsum(d / (n * (n - d)))
+
+        message = ("The variance estimate is undefined at some observations. "
+                   "This is feature of the mathematical formula.")
```
Nitpick to move that in the condition itself. Optional.

For the msg, I don't understand the formulation: `This is feature of`. Missing `a`?

mdhaber replied:
Yes, "a feature".
```suggestion
        message = ("The variance estimate is undefined at some observations. "
                   "This is a feature of the mathematical formula.")
```
Sometimes I intentionally take the message out of the branch. It documents what the branch is about.

mdhaber replied:
Corrected.

tupui commented on scipy/stats/_survival.py:
```diff
@@ -1,38 +1,213 @@
+        se = np.sqrt(var)
+        z = special.ndtri(1 / 2 + confidence_level / 2)
+
+        low = self.points - z * se
+        high = self.points + z * se
```
Nitpick, optional. Here and below.
```suggestion
        z_se = z * se
        low = self.points - z_se
        high = self.points + z_se
```

mdhaber replied:
Done.

tupui commented on scipy/stats/_survival.py:
```diff
@@ -1,38 +1,213 @@
+        z = special.ndtri(1 / 2 + confidence_level / 2)
+
+        lnl_points = np.log(-np.log(sf))
+        low = np.exp(-np.exp(lnl_points + z * se))
+        high = np.exp(-np.exp(lnl_points - z * se))
```
```suggestion
        z_se = z * se
        low = np.exp(-np.exp(lnl_points + z_se))
        high = np.exp(-np.exp(lnl_points - z_se))
```

mdhaber replied:
Done.

tupui commented on scipy/stats/_survival.py:
```diff
@@ -1,38 +1,213 @@
+            low, high = 1-high, 1-low
+        return low, high
+
+    def __call__(self, x):
```
As we discussed offline, an alternative would be to move this to `ECDFResult` instead. It could also take a name like `ECDFResult.evaluate`. This would return a subset of `ECDFResult` with just `sf` and `cdf` (possibly `x`). Both `sf` and `cdf` would still be `EmpiricalDistributionFunction` without the `__call__`.

Another option here if you wanted to keep the current API on `main` would be to add a `confidence_interval` method to `ECDFResult`. This would return two `ConfidenceInterval` (or another structure, the point being that it would hold both CI for CDF and SF.)

mdhaber replied:
Resolving because we decided to remove it for now.

tupui reviewed (commented):
Thanks Matt! Some suggestions

mdhaber commented on scipy/stats/tests/test_survival.py:
```diff
@@ -110,39 +153,110 @@ def test_nonunique(self):
+        assert_allclose(sf_ci.high, upper, atol=1e-5)
+        assert_allclose(cdf_ci.low, 1-upper, atol=1e-5)
+        assert_allclose(cdf_ci.high, 1-lower, atol=1e-5)
+
```
Optionally, add test against Mathematica for example 4.

@tupui, Mathematica's default confidence interval is the exponential greenwood.

```Mathematica
e = {24, 3, 11, 19, 24, 13, 14, 2, 18, 17, 24, 21, 12, 1, 10, 23, 6, 5, 9, 17}
ci = {1, 0, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 0, 1, 0, 1, 0, 1, 0}
R = EventData[e, ci]
S = SurvivalModelFit[R]
S["PointwiseIntervals", ConfidenceLevel->0.95, ConfidenceTransform->"LogLog"]
```

```{{0.694743,0.992802},{0.694743,0.992802},{0.647529,0.973299},{0.591142,0.947073},{0.591142,0.947073},{0.591142,0.947073},{0.591142,0.947073},{0.591142,0.947073},{0.591142,0.947073},{0.591142,0.947073},{0.464605,0.906422},{0.370359,0.856521},{0.370359,0.856521},{0.370359,0.856521},{0.370359,0.856521},{0.160489,0.776724},{0.160489,0.776724}} ```

mdhaber replied:
Done.

mdhaber commented on scipy/stats/tests/test_survival.py:
```diff
@@ -110,39 +153,110 @@ def test_nonunique(self):
+        assert_allclose(cdf_ci.low, 1-upper, atol=1e-5)
+        assert_allclose(cdf_ci.high, 1-lower, atol=1e-5)
+
+    def test_right_censored_ci_nans(self):
```
```Mathematica
e = {37, 43, 47, 56, 60, 62, 71, 77, 80, 81}
ci = {1, 1, 0, 0, 1, 1, 1, 0, 0, 0}
R = EventData[e, ci]
S = SurvivalModelFit[R]
S["PointwiseIntervals", ConfidenceLevel->0.95, ConfidenceTransform->"Linear"]
```

```
{{Indeterminate,Indeterminate},{Indeterminate,Indeterminate},{0.645828,1.10417},{0.449943,1.05006},{0.449943,1.05006},{0.449943,1.05006},{0.449943,1.05006},{0.0527015,0.947299},{-0.162389,0.662389},{Indeterminate,Indeterminate}} 
```

mdhaber replied:
So I'm not crazy: R's `survfit` does not think the first point is NaN:
```R
library(survival)
time = c(37, 43, 47, 56, 60, 62, 71, 77, 80, 81)
status = c(0, 0, 1, 1, 0, 0, 0, 1, 1, 1)
summary(survfit(Surv(time, status)~1, conf.type="plain", conf.int=0.95))
```
```
Call: survfit(formula = Surv(time, status) ~ 1, conf.type = "plain", 
    conf.int = 0.95)

 time n.risk n.event survival std.err lower 95% CI upper 95% CI
   47      8       1    0.875   0.117       0.6458        1.000
   56      7       1    0.750   0.153       0.4499        1.000
   77      3       1    0.500   0.228       0.0527        0.947
   80      2       1    0.250   0.210       0.0000        0.662
   81      1       1    0.000     NaN          NaN          NaN
```

mdhaber replied:
Saving results at other confidence levels for testing:
```
Call: survfit(formula = Surv(time, status) ~ 1, conf.type = "plain", 
    conf.int = 0.9)

 time n.risk n.event survival        std.err   lower 90% CI   upper 90% CI
   47      8       1    0.875 0.116926793337 0.682672539892 1.000000000000
   56      7       1    0.750 0.153093108924 0.498184244525 1.000000000000
   77      3       1    0.500 0.228217732294 0.124615235302 0.875384764698
   80      2       1    0.250 0.210406352883 0.000000000000 0.596087652672
   81      1       1    0.000            NaN            NaN            NaN
```
```
Call: survfit(formula = Surv(time, status) ~ 1, conf.type = "log-log", 
    conf.int = 0.9)

 time n.risk n.event survival        std.err    lower 90% CI   upper 90% CI
   47      8       1    0.875 0.116926793337 0.5002875589893 0.974583069009
   56      7       1    0.750 0.153093108924 0.3968379437078 0.914346087050
   77      3       1    0.500 0.228217732294 0.1290574025485 0.790844658509
   80      2       1    0.250 0.210406352883 0.0232126286119 0.600072521305
   81      1       1    0.000            NaN              NA             NA
```

mdhaber commented on scipy/stats/_survival.py:
```diff
@@ -1,38 +1,213 @@
+
+        method : str, optional
+            Method used to compute the confidence interval. Options are
+            "Greenwood" and "Exponential Greenwood".
```
Changed the names. Mathematica (and at least one R package) use "log-log" for "exponential Greenwood" and "linear" for "Greenwood".

mdhaber commented:
If tests are passing, I think this is about ready @tupui. I added tests that cover both types of confidence intervals at varying `confidence_levels` against a few references (R, Matlab, Mathematica), and I added a test that is mostly interested in NaNs that are produced by the formulas. There is a lot of disagreement here - we agree with R's `survival` package `survfit` function, and Mathematica and Matlab agree with each other, but there is some disagreement about leading NaNs that don't seem to be produced natively by the formulas. ISTM that lifelines turns NaNs into 1s or 0s. Anyway, I think we do a reasonable thing in following the formulas strictly, and R's `survival` package is solid precedent.

mdhaber commented:
To help us resolve lingering interface questions, I wanted to compile some information about existing nonparametric fit features in other libraries/languages. I'll write about them in the language of "durations" ("survival times", `x`) and "observed events" ("deaths", i.e. _not_ censored)

--

`lifelines` has a `KaplanMeierFitter` with a `fit` method that accepts the durations and booleans indicating whether the event was observed. The confidence level `alpha` must be provided when fitting, and there is only one method (log-log) for producing the confidence interval. All attributes we might want (point and confidence interval CDF/SF) are in a flat structure. The object has attributes:
* `durations` (our `x`)
* `cumulative_density_`, `survival_function_` (currently `cdf.points`, `sf.points) 
* `confidence_interval_survival_function_`, `confidence_interval_cumulative_density_`
* `median_survival_time_` (we considered having `ppf`, `isf` callables)
* `cumulative_density_at_times`, `survival_function_at_times` to evaluate CDF/SF at provided times
* `plot_cumulative_density`, `plot_survival_function`

It's nice to have a flat structure, but the downsides include a long list of attributes for a single object, long attribute names, and lots of documentation duplication. I also don't think that confidence limits should need to be provided at the times of fitting.

---

R's `survival` library has a `survfit` function that accepts a "formula" and returns a `survfit` object. It only computes the survival function and related quantities, not the CDF, so it's easier for it to be flat. The confidence level and method `conf.int`/`conf.type` are provided as additional arguments when calling `survfit`. All attributes we might want (point and confidence interval CDF/SF) are in a flat structure. It has attributes:
* `time` (our `x`)
* `surv` (our `sf.points`)
* `lower` and `upper` 

The downsides of this approach are similar to lifelines, and some of the bells and whistles we have considered don't seem to be implemented as attributes of the return object (although there might be other functions for doing these things). I'd rather keep this self-contained to avoid disorganized growth of SciPy's survival analysis features.

---

Matlab's [`ecdf`](https://www.mathworks.com/help/stats/ecdf.html) function is not a good reference for a Python API. For example, the function returns just points, points and confidence intervals, or nothing (but produces a plot) depending on how many variables there are to accept the output (e.g. `[f,x] = ecdf(y)` vs `[f,x,flo,fup] = ecdf(y)` vs `ecdf(y)`). You get the SF instead of the CDF by passing a pair of strings e.g. `[f,x] = ecdf(y, 'Function', 'survivor')`. Options for confidence interval level and type are specified similarly. This is not unusual for Matlab, but definitely not what we'd do in Python. 

---

Mathematica's [`SurvivalModelFit`](https://reference.wolfram.com/language/ref/SurvivalModelFit.html) is difficult to describe from a Python perspective. It returns an object with a lot of properties.
![image](https://user-images.githubusercontent.com/6570539/226217172-686ea3e0-0c81-473f-b800-ddb573c91606.png)

Some of those properties work like functions. For instance, `S["CDF"]` works like a function that can be evaluated at any point.
Some work like attributes. For instance, `S["PointwiseIntervals"]` returns confidence limits.
Sometimes the analogies break down. `S["PointwiseIntervals", ConfidenceLevel → 0.9, ConfidenceTransform → "LogLog"]` returns a 90% confidence interval with the log-log method. It can do a _lot_ but the syntax is very foreign.

tupui commented:
I think I am in general fine with the proposal. I just don't like much the plotting for the CI.

ATM we have a CI object which does not have a plotting semantic. If we add this capability here, then one could ask to do something similar for other CI objects (e.g. for Sobol' indices we can have arrays as well and could then ask to have a plot function.) Hence it would not be a namedTuple anymore but a normal object.

If we add a plot method, I would maybe prefer to directly be able to specify there a confidence level and have it plotted directly using this function (in this scenario `confidence_interval` would accept an input array to get CI for the given points and allow plotting values where we want.) Since SciPy is not really a plotting library, I am not sure we want to add too many plot functions.

mdhaber commented on scipy/stats/_survival.py:
```diff
@@ -1,38 +1,202 @@
+        self._sf = points if kind == 'sf' else 1 - points
+        self._kind = kind
+
+    def confidence_interval(self, confidence_level=0.95, method='linear'):
```
```suggestion
    def confidence_interval(self, confidence_level=0.95, *, method='linear'):
```

tupui commented on scipy/stats/_survival.py:
```diff
@@ -1,38 +1,202 @@
+        self._x = x
+        self._n = n
+        self._d = d
+        self._sf = points if kind == 'sf' else 1 - points
```
This is private so ok. If we have some follow up consider naming this `self._complement` so it's not tainted with CDF/SF.

tupui reviewed (approved):
Thanks Matt! LGTM with the latest updates. For people following, we had some extensive discussions offline about the API and I think we found a good compromise between simplicity, extendability and maintenance.

(I will merge after including the last suggestion about keyword only. The CI is green besides usual failures.)
